soccer/planning/planner/CollectPlanner.cpp

```diff
@@ -0,0 +1,139 @@
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius +
+                                  *_bufferDistBeforeContact);
+    Point noisyTargetVel =
+        approachDir * (ball.vel.dot(approachDir) + *_touchDeltaSpeed);
+    auto& targetInstant = _courseTargetInstants[request.shellID];
+    if (!targetInstant || targetInstant->pose.position().distTo(
+                              noisyTargetPoint) > *_targetChangeThreshold) {
+        targetInstant = RobotInstant{Pose{noisyTargetPoint, 0},
+                                     Twist{noisyTargetVel, 0}, RJ::Time{0s}};
+    }
+    request.motionCommand = PathTargetCommand{*targetInstant};
+    request.static_obstacles.add(
+        std::make_shared<Circle>(ball.pos, Ball_Radius));
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::fine(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    const Ball& ball = request.context->state.ball;
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    Point targetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius);
+    request.constraints.mot.maxSpeed =
+        ball.vel.dot(approachDir) + *_touchDeltaSpeed;
+    Point targetVel = request.constraints.mot.maxSpeed * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, Twist{targetVel, 0},
+                               RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::control(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    RobotInstant startInstant = request.start;
+    // move with constant negative acceleration until stop
+    double accel =
+        request.constraints.mot.maxAcceleration * *_controlAccelScalePercent;
+    double stoppingDist =
+        std::pow(startInstant.velocity.linear().mag(), 2) / (2 * accel);
+    Point targetPoint =
+        startInstant.pose.position() + stoppingDist * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, {}, RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    request.constraints.mot.maxSpeed = std::min(
+        startInstant.velocity.linear().mag(), request.constraints.mot.maxSpeed);
+    request.constraints.mot.maxAcceleration = accel;
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
```

Comment:
If the robot covers the ball during capture, the ball position will not update and will end up inside the robot. At that point, doing the subtraction will result in a direction behind the robot, resulting in a rapid spin that will kick the ball out